pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -163,3 +164,69 @@ def brick_sort(array, **kwargs):
 
     if _check_type(array, DynamicArray):
         array._modify(force=True)
+
+def _brick_sort_parallel_helper(array, start, end, comp, o):
+    is_sorted = True
+    for i in range(start+o, end, 2):
+        if _comp(array[i+1], array[i], comp):
+            array[i], array[i+1] = array[i+1], array[i]
+            is_sorted = False
+    return is_sorted
+
+def brick_sort_parallel(array, **kwargs):
+    """
+    Implements Concurrent Brick Sort / Odd Even sorting algorithm
+
+    Parameters
+    ==========
+
+    array: Array
+        The array which is to be sorted.
+    start: int
+        The starting index of the portion
+        which is to be sorted.
+        Optional, by default 0
+    end: int
+        The ending index of the portion which
+        is to be sorted.
+        Optional, by default the index
+        of the last position filled.
+    comp: lambda/function
+        The comparator which is to be used
+        for sorting. If the function returns
+        False then only swapping is performed.
+        Optional, by default, less than or
+        equal to is used for comparing two
+        values.
+
+    Examples
+    ========
+    >>> from pydatastructs import OneDimensionalArray, brick_sort_parallel
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> brick_sort_parallel(arr)
+    >>> [arr[0], arr[1], arr[2]]
+    [1, 2, 3]
+    >>> brick_sort_parallel(arr, comp=lambda u, v: u > v)
+    >>> [arr[0], arr[1], arr[2]]
+    [3, 2, 1]
+
+    References
+    ==========
+    .. [1] https://www.geeksforgeeks.org/odd-even-sort-brick-sort/
+    """
+
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+
+    is_sorted = False
+    while is_sorted is False:
+        is_sorted = True
+        with ThreadPoolExecutor(max_workers=2) as executor:
+            t1 = executor.submit(_brick_sort_parallel_helper, array, start, end, comp, 1)
+            t2 = executor.submit(_brick_sort_parallel_helper, array, start, end, comp, 0)
+
+            is_sorted = t1.result() and t2.result()
```

Comment:
A better way to parallelize brick sort is two compare and swap elements on multiple threads. The current implementation leads too many races(which may be benign but reduce the efficiency).
See, [this image](https://www.google.com/url?sa=i&url=http%3A%2F%2Fwww.darshan.ac.in%2FUpload%2FDIET%2FDocuments%2FCE%2F180702_PP%2520-%2520IMP%2520Questions%2520and%2520Solution%2520-%25202015_17042015_061013AM.pdf&psig=AOvVaw1jFpyWZvus_IqgOe19ZdSh&ust=1585406147731000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCNjco63wuugCFQAAAAAdAAAAABAI) to understand my point.

Reply:
the function which has i = 1 will work on odd indexes and i = 0 will work on even indexes. The link is sending me to a pdf, shall I read it whole or is there a specific page number you want me to refer?

Reply:
![download](https://user-images.githubusercontent.com/36567889/77770634-97464d80-706b-11ea-9f9e-3fa08376de48.png)
See here, various pair of indices can be parallely compared independently without any race in a specific phase. So, after running even number of phases we can check through `is_sort` whether the array is sorted or not.

Reply:
ok correct me if I am wrong, you are saying that to assign each swap in Phase 1 its own thread which compares and swaps the their 2 indexes and when phase 1 is completed then phase 2 should start. So concurrency in the Phase itself but not between the 2 phases?

Reply:
Yes, like if we are using some `num_threads` number of threads then the swaps can be distributed among those threads. Though the races are benign if two threads start from different indices but having two threads instead of one isn't really that useful. The time will be reduced by at most half.

Reply:
I really like the idea, will make the required changes by tomorrow.

Reply:
I have made the required changes, please check it out